client/src/pages/Challenge.tsx

```diff
@@ -1,118 +1,181 @@
+  Container,
+  Row,
+  Col,
+  ButtonToolbar,
+  Button,
+  AccordionItem,
+  AccordionHeader,
+  AccordionBody,
+  Accordion,
+} from "react-bootstrap";
 import { useCallback, useEffect, useState } from "react";
 import { useParams } from "react-router-dom";
 import { ChallengeDetails } from "../types/challengeDetails";
 import { StarFill } from "react-bootstrap-icons";
 
-
 const Challenge = () => {
-    //TODO: Fetch this value from the server
-    const [completed, setCompleted] = useState(false); //TODO: This value should be fetched from the server
-    const [details, setDetails] = useState<ChallengeDetails>();
-    const [isLoading, setIsLoading] = useState(true);
+  //TODO: Fetch this value from the server
+  const [completed, setCompleted] = useState(false);
+  const [details, setDetails] = useState<ChallengeDetails>();
+  const [isLoading, setIsLoading] = useState(true);
 
-    const { id } = useParams();
+  const { id } = useParams();
 
+  useEffect(() => {
+    //fetch data about the challenge with the provided "id"
+    fetch("/api/challenges/" + id)
+      .then((response) => {
+        if (!response.ok) {
+          throw new Error(response.statusText);
+        }
+        return response.json() as Promise<ChallengeDetails>;
+      })
+      .then((data) => {
+        setDetails(data);
+        setIsLoading(false);
+        // console.log(details);
+        return;
+      })
+      .catch((err) => {
+        console.log();
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+        console.error(
+          "Failed fetching the challenge number: " +
+            id +
+            "\nError message: " +
+            err.message,
+        );
+      });
+  }, [id]);
 
-    useEffect(() => {   
-        //fetch data about the challenge with the provided "id"
-        fetch('/api/challenges/' + id).then((response) => {
-            if(!response.ok){
-                throw new Error(response.statusText);                
-            }
-            return response.json() as Promise<ChallengeDetails>;
-        }).then((data) => {
-            setDetails(data);
-            setIsLoading(false);
-            // console.log(details);
-            return;
-        }).catch((err) => {
-            console.log();
-            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
-            console.error("Failed fetching the challenge number: " + id + "\nError message: " + err.message)
-        });
-   }, [id]);
+  // TODO: optional - can be converted to useMemo, or even removed until performance
+  // becomes an issue.
 
-    const getDifficultyStars = useCallback((difficulty: number) => {
-        return Array.from({ length: difficulty + 1 }, (_, i) => <StarFill key={i} />);
-    }, [details]);
+  const getDifficultyStars = useCallback(
+    (difficulty: number) => {
+      return Array.from({ length: difficulty + 1 }, (_, i) => (
+        <StarFill key={i} />
+      ));
+    },
+    [details],
```

Comment:
How does `getDifficultyStars` does not seem to depend on `details`?

Also why does the function generate `difficulty + 1 `stars? Is this one-off intentional ?

Reply:
Same problem as before, details is not in the function definition, what @Vladmidir intended to do here as I see it is that since difficulty is a part of details, the function should change when details changes. Again, refactor with useMemo and details as dependency.

Also difficulty varies from 0 - 2 and star count should vary from 1 - 3, so it's intentional.